File: en/codes/rust/chapter_hashing/hash_map_chaining.rs
/*
 * File: hash_map_chaining.rs
 * Created Time: 2023-07-07
 * Author: WSL0809 (wslzzy@outlook.com)
 */

#[derive(Clone)]
/* Key-value pair */
struct Pair {
    key: i32,
    val: String,
}

/* Hash table with separate chaining */
struct HashMapChaining {
    size: usize,
    capacity: usize,
    load_thres: f32,
    extend_ratio: usize,
    buckets: Vec<Vec<Pair>>,
}

impl HashMapChaining {
    /* Constructor */
    fn new() -> Self {
        Self {
            size: 0,
            capacity: 4,
            load_thres: 2.0 / 3.0,
            extend_ratio: 2,
            buckets: vec![vec![]; 4],
        }
    }

    /* Hash function */
    fn hash_func(&self, key: i32) -> usize {
        key as usize % self.capacity
    }

    /* Load factor */
    fn load_factor(&self) -> f32 {
        self.size as f32 / self.capacity as f32
    }

    /* Remove operation */
    fn remove(&mut self, key: i32) -> Option<String> {
        let index = self.hash_func(key);

        // Traverse bucket and remove key-value pair from it
        for (i, p) in self.buckets[index].iter_mut().enumerate() {
            if p.key == key {
                let pair = self.buckets[index].remove(i);
                self.size -= 1;
                return Some(pair.val);
            }
        }

        // If key is not found, return None
        None
    }

    /* Expand hash table */
    fn extend(&mut self) {
        // Temporarily store the original hash table
        let buckets_tmp = std::mem::take(&mut self.buckets);

        // Initialize expanded new hash table
        self.capacity *= self.extend_ratio;
        self.buckets = vec![Vec::new(); self.capacity as usize];
        self.size = 0;

        // Move key-value pairs from original hash table to new hash table
        for bucket in buckets_tmp {
            for pair in bucket {
                self.put(pair.key, pair.val);
            }
        }
    }

    /* Print hash table */
    fn print(&self) {
        for bucket in &self.buckets {
            let mut res = Vec::new();
            for pair in bucket {
                res.push(format!("{} -> {}", pair.key, pair.val));
            }
            println!("{:?}", res);
        }
    }

    /* Add operation */
    fn put(&mut self, key: i32, val: String) {
        // When load factor exceeds threshold, perform expansion
        if self.load_factor() > self.load_thres {
            self.extend();
        }

        let index = self.hash_func(key);

        // Traverse bucket, if specified key is encountered, update corresponding val and return
        for pair in self.buckets[index].iter_mut() {
            if pair.key == key {
                pair.val = val;
                return;
            }
        }

        // If key does not exist, append key-value pair to the end
        let pair = Pair { key, val };
        self.buckets[index].push(pair);
        self.size += 1;
    }

    /* Query operation */
    fn get(&self, key: i32) -> Option<&str> {
        let index = self.hash_func(key);

        // Traverse bucket, if key is found, return corresponding val
        for pair in self.buckets[index].iter() {
            if pair.key == key {
                return Some(&pair.val);
            }
        }

        // If key is not found, return None
        None
    }
}

/* Driver Code */
pub fn main() {
    /* Initialize hash table */
    let mut map = HashMapChaining::new();

    /* Add operation */
    // Add key-value pair (key, value) to the hash table
    map.put(12836, "Xiao Ha".to_string());
    map.put(15937, "Xiao Luo".to_string());
    map.put(16750, "Xiao Suan".to_string());
    map.put(13276, "Xiao Fa".to_string());
    map.put(10583, "Xiao Ya".to_string());
    println!("\nAfter adding is complete, hash table is\nKey -> Value");
    map.print();

    /* Query operation */
    // Input key into hash table to get value
    println!(
        "\nInput student ID 13276, found name {}",
        match map.get(13276) {
            Some(value) => value,
            None => "Not a valid Key",
        }
    );

    /* Remove operation */
    // Remove key-value pair (key, value) from hash table
    map.remove(12836);
    println!("\nAfter removing 12836, hash table is\nKey -> Value");
    map.print();
}

File: ja/codes/go/chapter_hashing/hash_map_chaining.go
// File: hash_map_chaining.go
// Created Time: 2023-06-23
// Author: Reanon (793584285@qq.com)

package chapter_hashing

import (
	"fmt"
	"strconv"
	"strings"
)

/* チェイン法ハッシュテーブル */
type hashMapChaining struct {
	size        int      // キーと値のペア数
	capacity    int      // ハッシュテーブル容量
	loadThres   float64  // リサイズを発動する負荷率のしきい値
	extendRatio int      // 拡張倍率
	buckets     [][]pair // バケット配列
}

/* コンストラクタ */
func newHashMapChaining() *hashMapChaining {
	buckets := make([][]pair, 4)
	for i := 0; i < 4; i++ {
		buckets[i] = make([]pair, 0)
	}
	return &hashMapChaining{
		size:        0,
		capacity:    4,
		loadThres:   2.0 / 3.0,
		extendRatio: 2,
		buckets:     buckets,
	}
}

/* ハッシュ関数 */
func (m *hashMapChaining) hashFunc(key int) int {
	return key % m.capacity
}

/* 負荷率 */
func (m *hashMapChaining) loadFactor() float64 {
	return float64(m.size) / float64(m.capacity)
}

/* 検索操作 */
func (m *hashMapChaining) get(key int) string {
	idx := m.hashFunc(key)
	bucket := m.buckets[idx]
	// バケットを走査し、key が見つかれば対応する val を返す
	for _, p := range bucket {
		if p.key == key {
			return p.val
		}
	}
	// key が見つからない場合は空文字列を返す
	return ""
}

/* 追加操作 */
func (m *hashMapChaining) put(key int, val string) {
	// 負荷率がしきい値を超えたら、リサイズを実行
	if m.loadFactor() > m.loadThres {
		m.extend()
	}
	idx := m.hashFunc(key)
	// バケットを走査し、指定した key が見つかれば対応する val を更新して返す
	for i := range m.buckets[idx] {
		if m.buckets[idx][i].key == key {
			m.buckets[idx][i].val = val
			return
		}
	}
	// その key が存在しなければ、キーと値のペアを末尾に追加
	p := pair{
		key: key,
		val: val,
	}
	m.buckets[idx] = append(m.buckets[idx], p)
	m.size += 1
}

/* 削除操作 */
func (m *hashMapChaining) remove(key int) {
	idx := m.hashFunc(key)
	// バケットを走査してキーと値のペアを削除
	for i, p := range m.buckets[idx] {
		if p.key == key {
			// スライスから削除する
			m.buckets[idx] = append(m.buckets[idx][:i], m.buckets[idx][i+1:]...)
			m.size -= 1
			break
		}
	}
}

/* ハッシュテーブルを拡張 */
func (m *hashMapChaining) extend() {
	// 元のハッシュテーブルを一時保存
	tmpBuckets := make([][]pair, len(m.buckets))
	for i := 0; i < len(m.buckets); i++ {
		tmpBuckets[i] = make([]pair, len(m.buckets[i]))
		copy(tmpBuckets[i], m.buckets[i])
	}
	// リサイズ後の新しいハッシュテーブルを初期化
	m.capacity *= m.extendRatio
	m.buckets = make([][]pair, m.capacity)
	for i := 0; i < m.capacity; i++ {
		m.buckets[i] = make([]pair, 0)
	}
	m.size = 0
	// キーと値のペアを元のハッシュテーブルから新しいハッシュテーブルへ移す
	for _, bucket := range tmpBuckets {
		for _, p := range bucket {
			m.put(p.key, p.val)
		}
	}
}

/* ハッシュテーブルを出力 */
func (m *hashMapChaining) print() {
	var builder strings.Builder

	for _, bucket := range m.buckets {
		builder.WriteString("[")
		for _, p := range bucket {
			builder.WriteString(strconv.Itoa(p.key) + " -> " + p.val + " ")
		}
		builder.WriteString("]")
		fmt.Println(builder.String())
		builder.Reset()
	}
}
